### ISSUE TITLE
remult Next.js Tutorial

### ISSUE THREAD

**Issue by quangtynu** [remult/remult#145]
Now i go to "Refactor from Front-end to Back-end" of Tutorial remult for Next.js at link:
https://remult.dev/tutorials/react-next/backend-methods.html#set-all-tasks-as-un-completed
when npm run dev: error is folow in console:

{
  message: "can't use static remult in this environment, `async_hooks` were not initialized",
  stack: [
    "Error: can't use static remult in this environment, `async_hooks` were not initialized",
    '    at RemultAsyncLocalStorage.getRemult (/home/quangtynu/remult-nextjs-app/node_modules/remult/server/expressBridge.js:114:19)',
    '    at RemultProxy.remult_proxy_1.remult.remultFactory (/home/quangtynu/remult-nextjs-app/node_modules/remult/server/expressBridge.js:96:54)',
    '    at RemultProxy.repo (/home/quangtynu/remult-nextjs-app/node_modules/remult/src/remult-proxy.js:20:32)',
    '    at setAll (webpack-internal:///(api)/./src/shared/TasksController.ts:19:69)',
    '    at myServerAction.originalMethod (/home/quangtynu/remult-nextjs-app/node_modules/remult/src/server-action.js:197:114)',
    '    at myServerAction.<anonymous> (/home/quangtynu/remult-nextjs-app/node_modules/remult/src/server-action.js:146:71)',
    '    at step (/home/quangtynu/remult-nextjs-app/node_modules/tslib/tslib.js:144:27)',
    '    at Object.next (/home/quangtynu/remult-nextjs-app/node_modules/tslib/tslib.js:125:57)',
    '    at fulfilled (/home/quangtynu/remult-nextjs-app/node_modules/tslib/tslib.js:115:62)',
    '    at processTicksAndRejections (node:internal/process/task_queues:96:5)'
  ],
  url: '/api/setAll',
  method: 'POST'
}

I using Ubuntu on WSL2.
any idea about async_hooks?

**Comment by noam-honig** [comment]
Hi @quangtynu,

Great catch - I see the issue and am working on it.

In the meantime, let me suggest a workaround, please add an optional parameter called `remult` of type `Remult` to the `setAll` method. (, remult?:Remult)
And adjust the code that defined the repo by adding a '!' before the `.repo`. should be  `remult!.repo(Task)`

You don't need to send that parameter when you call it - the `BackendMethod` decorator'll automatically inject it.

So your code should look like this:
```ts
import { BackendMethod, Remult } from "remult";
import { Task } from "./Task";

export class TasksController {
  @BackendMethod({ allowed: true })
  static async setAll(completed: boolean, remult?: Remult) {
    const taskRepo = remult!.repo(Task);
    for (const task of await taskRepo.find()) {
      await taskRepo.save({ ...task, completed });
    }
  }
}
```
You can see a working example at a code sandbox I've created at:
https://codesandbox.io/p/sandbox/interesting-feistel-poit6u?file=%2Fsrc%2Fshared%2FTasksController.ts

Of course, I'm still working on the original issue you reported.

Please let me know that the workaround also works for you

**Comment by quangtynu** [comment]
Ok, I found out:
modified code in file "src/shared/TasksController.ts" 👍 
Origin code as described in:
https://remult.dev/tutorials/react-next/backend-methods.html#refactor-from-front-end-to-back-end

import { BackendMethod, remult } from "remult";
import { Task } from "./Task";

export class TasksController {
   @BackendMethod({ allowed: true })
   static async setAll(completed: boolean) {
      const taskRepo = remult.repo(Task);

      for (const task of await taskRepo.find()) {
         await taskRepo.save({ ...task, completed });
      }
   }
}
//---------------------------------
to this:

import { BackendMethod, Remult } from "remult";
import { Task } from "./Task";

export class TasksController {
   @BackendMethod({ allowed: true })
   static async setAll(completed: boolean, remult?: Remult) {
      const taskRepo = remult!.repo(Task);

      for (const task of await taskRepo.find()) {
         await taskRepo.save({ ...task, completed });
      }
   }
}

//---------------------------------
 as described in:
https://remult.dev/docs/backendMethods.html#static-backend-method
it works now. But i dont interstand.

**Comment by noam-honig** [comment]
@quangtynu, thanks for letting me know that it works.

Let me try and clarify.

When using a `BackendMethod` you can define an optional parameter of type `Remult` and the `BackendMethod` will automatically inject the correct `remult` object in the context of the incoming request.

Actually - before version 0.16, this was the only way to use `remult` on a backend method.

In version 0.16 we introduced the ability to just use the `remult` object directly without defining the optional parameter - we did it by using the `async_hooks` feature of node.

Since not all javascript environment support `async_hooks` (for example, demo) - we made it optional.

In your case - in the `next` tutorial - the instructions at this time do not include an explanation on how to enable that feature that depends on `async_hooks`.

So - while we fix the tutorial, I wanted to give you a quick workaround.

To summarize - for now (until we fix the tutorial), when you use backend methods, add an optional parameter of type `Remult` to gain access to the `remult` object.

Once we update the tutorial, I'll let you know.

Sounds good?

**Comment by quangtynu** [comment]
Thank you for Your explain. it nice for me now. actually I also try to find into how its works.

**Comment by quangtynu** [comment]
Now i come to: "Frontend setup" of "Authentication and Authorization" part:
https://remult.dev/tutorials/react-next/auth.html#user-authentication

the error in "pages/index.tsx" is on the browser:

Server Error
Error: [next-auth]: `useSession` must be wrapped in a <SessionProvider />

This error happened while generating the page. Any console logs will be displayed in the terminal window.
Source
pages/index.tsx (19:38) @ Home

  17 | 
  18 | const Home: NextPage = () => {
> 19 | const { data: session } = useSession();
     |                                    ^
  20 | 
  21 | const [tasks, setTasks] = useState<Task[]>([]);
  22 | const [hideCompleted, setHideCompleted] = useState(false);
Show collapsed frames

**Comment by yoni-rapoport** [comment]
It seems that in the current version of next-auth you have to add a <SessionProvider> to the _app.tsx for useSession() to work.

https://next-auth.js.org/getting-started/client#sessionprovider

We need to fix the tutorial to reflect that. Thanks!

**Comment by quangtynu** [comment]
Thank you. @noam-honig . It works now.
